그룹 도메인 로직을 추가합니다

agvber opened the pull request:
## 🤷‍♂️ PR 내용(Issue 꼭 달기!)
- User Model 코드 리펙토링
- User Model - group id 추가
- 가족 그룹 생성 use case 추가
- 가족 그룹 참가 use case 추가

## 설명

### CreateFamilyGroupUseCase 가족 그룹 생성 use case
1. 그룹을 생성
2. 그룹 초대 코드를 생성
3. InvitationMessage 결과를 리턴

```
data class InvitationMessage(
    val url: String,
    val code: String,
)
```
URL -> 해당 그룹에 초대 URL 링크이며 초대 할 대상에게 전달 하여야 합니다.
CODE -> 초대 받은 URL로 접속하는 경우 초대 코드와 함께 작성해서 그룹을 참여할 수 있습니다.



### JoinFamilyGroupUseCase 가족 그룹 참여 use case
1. 가족 그룹을 참여 합니다

```
data class InvitedGroup(
    val groupId: Int,
    val groupManagerInfo: GroupManagerInfo,
    val users: List<User>
) {

    data class GroupManagerInfo(
        val name: String
    )

    data class User(
        val imageUrl: String?
    )
}
```

GroupManagerInfo -> 해당 그룹의 매니저 정보이며 [**손녀 조다은님**] 해당 그룹의 매니저 정보를 담고 있습니다.
User -> 해당 그룹을 이미 가입한 사람들의 정보이며 프로필 이미지 값을 리스트 형태로 반환 합니다.

<img src="https://github.com/user-attachments/assets/964eca33-a060-4210-b86e-1c591b89ee61" width=30% height=30%/>

agvber committed:
feat: 로그인 필요 여부를 판단하는 usecase 추가

agvber committed:
Merge remote-tracking branch 'origin/develop' into feature/auth

agvber committed:
refactor: emptyUser 변수 제거

agvber committed:
feat: User Model group_id 추가

agvber committed:
feat: GroupManagementRepository 추가

agvber committed:
feat: 가족 그룹 생성 usecase 추가

agvber committed:
feat: 가족 그룹 참가 usecase 추가

DJDrama commented on core/data/src/main/java/com/teampatch/core/data/repository/GroupManagementRepositoryImpl.kt:
```diff
@@ -0,0 +1,48 @@
+class GroupManagementRepositoryImpl @Inject constructor(
+    private val tokenManager: TokenManager,
+    private val groupRemoteDataSource: GroupRemoteDataSource,
+    private val getUserInfoUseCase: GetUserInfoUseCase
```
UseCase는 중재자 역할을 하는 UI와 Data 영역을 이어주는, 중재해주는 역할을 합니다.
따라서 UseCase는 repository에 주입되면 clean architecture 규칙을 위반하게 됩니다~!
따라서 getUserInfoUserCase가 호출하는 Repository의 dataSource를 주입하면 좋을 것 같습니다!

agvber replied:
해당 부분의 개념이 부족했던 것 같았네요 감사합니다!!


아래와 같이 (Domain)Repository interface를 의존하는 것은 문제없을 것 같다고 생각되는데 
어떻게 생각하시나요??
```
class GroupManagementRepositoryImpl @Inject constructor(
    private val tokenManager: TokenManager,
    private val groupRemoteDataSource: GroupRemoteDataSource,
    private val userRepository: UserRepository
) : GroupManagementRepository {
}
```

DJDrama replied:
네 좋습니다~!

DJDrama replied:
조금 더 정확하게 하려면 repository를 repository에 주입하지 말구 datasource로만 주입하면 좋을 것 같습니다. 왜냐하면 repository끼리는 동등한 레벨이라.

agvber replied:
흠 아직 감이 안 와서 고민해보겠습니다!

DJDrama commented on core/domain/src/main/java/com/teampatch/core/domain/model/User.kt:
```diff
@@ -2,8 +2,21 @@ package com.teampatch.core.domain.model
+) {
+    companion object {
+
+        fun init(): User = User(
```
init() 보다는 createEmptyUser 같은 명명법이 좋아보입니다~!
왜냐면 실질적으로 init {} 블록이랑 헷갈릴 수 있기 때문이죠!

agvber replied:
해당 부분까지 생각 하지 못 했네요! 수정했습니다 🙂

DJDrama commented on core/domain/src/main/java/com/teampatch/core/domain/usecase/group/CreateFamilyGroupUseCase.kt:
```diff
@@ -0,0 +1,16 @@
+
+    suspend operator fun invoke(): InvitationMessage {
+        val invitationUrl = groupManagementRepository.createFamilyGroup()
+        val invitationCode = groupManagementRepository.generateInviteCode()
```
이럴 때는 coroutineScope을 이용해서 async(deferred) await을 사용해보면 좋을 것 같습니다~!

agvber replied:
혹시 병렬 처리 부분을 말씀하시는 것일까요?? (틀렸다면 넘어가 주시면 될 것 같습니다ㅠㅠ)

그룹을 생성 한 후에 그룹 초대 코드를 만들 수 있어서 병렬 처리를 하지 못 하였습니다...

1. groupManagementRepository.createFamilyGroup() -> 그룹을 생성한다.
2. groupManagementRepository.generateInviteCode() -> 그룹 초대 코드를 생성한다.

DJDrama replied:
아하 그렇군요~! 그럼 괜찮을 것 같습니다.

agvber committed:
refactor: GroupManagementRepositoryImpl 구현체usecase 의존성 제거 https://github.com/Team-PATCH/Harmony-Android/pull/76#discussion_r1894885303

agvber committed:
refactor: TokenManager Hilt 인스턴스 생성 시점 변경 (ViewModel -> Application)

agvber committed:
fix: user flow 에러가 catch 되지 않는 문제 해결

[설명]
flow 바깥에 있는 로직이 error가 발생할 때 flow catch로 잡히지 않아,
runCatching을 사용하였지만 flow 안에 있는 로직이 에러가 나는 경우 잡히지 않아 새로운 flowErrorCatch 함수를 작성함.

agvber committed:
feat: PagingData 에러 핸들링 utils 함수 추가

agvber committed:
refactor: HomeViewModel 에러 핸들링 리펙토링

agvber committed:
Merge remote-tracking branch 'origin/develop' into feature/group_domain

# Conflicts:
#	core/common/src/main/java/com/teampatch/core/common/Paging.kt
#	core/data/src/main/java/com/teampatch/core/data/di/DataSingletonModule.kt
#	core/data/src/main/java/com/teampatch/core/data/di/DataViewModelModule.kt
#	core/domain/src/main/java/com/teampatch/core/domain/model/User.kt
#	core/domain/src/main/java/com/teampatch/core/domain/usecase/user/GetUserInfoUseCase.kt

agvber committed:
style: 코드 포멧 스타일 정리
